Feature/add new breed

gfrantzsword opened the pull request:
## :bookmark:	Ticket Link 

N/A

## :interrobang: Type of change

:new: New feature

## :mag_right: How to validate the PR

#### (If you haven't done so yet)
Add your Cat API key to `local.properties`:
   ```
   API_KEY="your_cat_api_key_here"
   ```

#### Validate

1. Open the `Add a new breed` modal (`+`) from the `Breeds List` main screen
2. Fill in the input fields and submit
3. Verify the updated list

## :open_book: Description

Added a feature to create new breeds

## :camera_flash:	Screenshots

New 'add' button in the top right
<img width="270" height="600" alt="image" src="https://github.com/user-attachments/assets/49404829-265d-4977-8862-68ff64869a45" />

'Add a new breed' modal
<img width="270" height="600" alt="image" src="https://github.com/user-attachments/assets/0aac8eb8-94e0-4fc9-81c2-377c730d1b30" />

## :white_check_mark: Checklist

- [x] I have created/updated unit tests
- [x] New and existing unit tests pass locally with my changes
- [ ] I have created documentation

gfrantzsword committed:
feature: Initial structure for new-breed dialog

- Created ModalBottomSheet and NewBreedTextField composable;
- Roughly organized the form layout on the bottom sheet;
- Updated dimensions to add new spacing category

gfrantzsword committed:
fix: Correct color on drag section of ModalBottomSheet

- Add containerColor to match the content background

gfrantzsword committed:
feature: Created new temperamentChips section on new-breed form

- Created a new FlowRow to display the selectable chips with expand functionality
- Added new values to Dimensions.kt

gfrantzsword committed:
fix: Small adjustments to NewBreedSheet

- Made the NewBreedSheetContent scrollable;
- Added a Spacer to pin the submit button to the bottom of the screen;
- Updated the 'show more' button color and added indicative '+'/'-'
- Changed the dismiss button from a navigationIcon to an action (now displayed on the right instead of the left)

gfrantzsword committed:
feature: Turn the 'origin' TextField into a Dynamic Dropdown

- Created NewBreedDropDwonTextField with an ExposedDropdownMenuBox to display existing items based on user input
- Updated BreedListViewModel to filter origins to be displayed

gfrantzsword committed:
feature: Prepared ViewModel and NewBreedTextField to support error messages

- Added isError, errorMessage and supportingText to NewBreedTextField
- Filtered 'allNames' on ViewModel for text field verifications

gfrantzsword committed:
Merge branch 'ui-general-update' into feature/add_new_breed

# Conflicts:
#	core/src/main/java/com/example/catbreeds/core/ui/theme/Dimensions.kt
#	features/breed_list/src/main/java/com/example/catbreeds/breed_list/BreedListScreen.kt

gfrantzsword committed:
refactor: merged 'ui-general.update' branch + adapted changes

- Updated BreedList's Screen and ViewModel to consider the changes on Breed model;
- Applied suggestions to fit the team's approach

gfrantzsword committed:
feature: Add new description Text Field

- Added new multiLine description Text Field;
- Updated instances to use constant `DefaultWeight`

gfrantzsword committed:
feature: Add new `MaxCharCount` on Text Fields

- Added supporting text to display current and max char count
- Limited input size to `MaxCharCount`
- Defined constants on Dimensions.kt

gfrantzsword committed:
feature: Add field and submit validation

- Specific error messages for 'repeated name' or 'Max<Min' in real-time;
- Required field validation on submit;
- Colored warning messages and labels

gfrantzsword committed:
feature: Add title to temperaments and life-expectancy fields

gfrantzsword committed:
refactor: Update temperaments section to use inline expand button

- "See more" button is now of type `SelectTemperamentBreed` for seamless appearance

gfrantzsword committed:
refactor: Animate NewBreedSheet dismissal

- Waits for the slide down animation to complete before dismissing the composable

gfrantzsword committed:
feature: Add 'unsaved changes' alert

- Add 'dirty' state tracker;
- Intercept bottom sheet close and display alert dialog if there are unsaved changes;
- Use `onSuccess` for successful submission

gfrantzsword committed:
Refactor: Updated 'remember' definition to fit team's approach

from:
```
val ... by remember
```

to:
```
val ... = remember
```

gfrantzsword committed:
fix: Added missing parameter to fix verification error message not appearing

- Added `isMaxLessThanMin` parameter on `maxLifeError`

gfrantzsword committed:
refactor: Moved 'modifier' parameter on NewBreedTextField methods to first position

gfrantzsword committed:
Merge branch 'ui-general-update' into feature/add_new_breed

gfrantzsword committed:
feature: Implement `AddNewBreed` functionality

- Created `addBreed(breed: Breed)` on repository and `insert(breed: BreedEntity)` and DAO to save new breeds locally.
- Added `onSucess` logic to immediately open the new breed's detail screen
- Defined default `imageUrl` and `isFavorite` to simplify future implementation on form

gfrantzsword committed:
Merge branch 'ui-general-update' into feature/add_new_breed

gfrantzsword committed:
test: Create new `addNewBreed` test

Create new test for addNewBreed feature on RepositoryTest.kt

gfrantzsword committed:
fix: Fixed oversized bottom padding on NewBreedSheetContent

- Updated the order of the Column modifiers that were causing an oversized bottom padding noticeable when scrolling

gfrantzsword committed:
feature: Implemented image field with gallery selection on `addNewBreed`

- Created `NewBreedImageField`
- Added logic to copy imageUri` to local storage
- Updated dimensions.kt

The logic to have an image on the new breed was already mostly implemented, which made the changes much less complex.

gfrantzsword committed:
fix: Correctly import `android.net.Uri` on `BreedListScreen.kt`

gfrantzsword committed:
feature: Added camera support on image field

- Added camera `<uses-feature/>` on AndroidManifest.kt
- Configured FileProvider and file_paths.xml
- Updated BreedListScreen to include a gallery-camera option dialog

gfrantzsword committed:
refactor: Update NewBreedImageField with new UI

- Replaced alertDialog with permanent options
- Updated placeholder image to use `ic_cat_placeholder` drawable
- Overall more coherent design

gfrantzsword committed:
refactor: Create `NewBreedSheet.kt` file

- MOved the `NewBreedSheetContent` composable and helper methods into a dedicated file

gfrantzsword committed:
refactor: Reorganized `NewBreedSheet.kt` for modularity and readability

- Extracted validation logic from submit button
- Created specific `LifeExpectancyField`, `TemperamentField` and `OriginField` for unique field types
- Overall formatting and comments

gfrantzsword committed:
feature: Created new `FieldTitle` composable

Created new composable with title and subtitle to be used on multiple fields

gfrantzsword committed:
refactor: Replace submit `Button()` with `ExtendedFloatingActionButton()`

gfrantzsword committed:
refactor: Update `breed.lifeSpan` display on details and remove from favorites

- Added indicative 'years' to `{min} - {max}` display on details screen
- Removed lifeSpan display from favorites screen

gfrantzsword committed:
feature: Reload list breeds on resume

- Make `refreshBreeds` public
- Add `LifecycleOwner` observer
- Add necessary dependency

This ensures the list maintains a consistent order after the addition of new breeds

gfrantzsword committed:
fix: Add `context` to `breedListViewModelTest.kt`

- Fixed compilation error

goncalo-guilherme-sword commented on core/src/main/java/com/example/catbreeds/core/ui/theme/Dimensions.kt:
```diff
@@ -6,15 +6,23 @@ private object BaseValues {
 
     val ShadowSmall = 6.dp
 
+    val ConstNumberSmall = 6
```
Consider using the Dimensions class only to have values that are used for dimensions (width, height, size, padding, ...).
For this new values, maybe you could consider having a Constants file where you can put them there.

gfrantzsword committed:
fix: Update DB schema

goncalo-guilherme-sword commented on features/breed_list/src/main/java/com/example/catbreeds/breed_list/BreedListViewModel.kt:
```diff
@@ -125,4 +163,64 @@ class BreedListViewModel @Inject constructor(
         _errorMessage.value = null
     }
+
+    fun addNewBreed(
```
In this method, maybe you can use the Breed class to encapsulate the information about the new breed you want to add. What do you think?

gfrantzsword committed:
refactor: Separate constants from `Dimensions.kt` into dedicated file

AlexCarvalhoSword commented on features/breed_list/src/main/java/com/example/catbreeds/breed_list/BreedListScreen.kt:
```diff
@@ -25,37 +58,77 @@ import androidx.compose.ui.focus.FocusRequester
+    DisposableEffect(lifecycleOwner) {
+        val observer = LifecycleEventObserver { _, event ->
+            if (event == Lifecycle.Event.ON_RESUME) {
+                viewModel.refreshBreeds()
```
with this change isn't the refreshBreeds() called twice ? Here and on the BreedListViewModel's init

AlexCarvalhoSword commented on features/breed_list/src/main/java/com/example/catbreeds/breed_list/BreedListScreen.kt:
```diff
@@ -83,6 +156,74 @@ fun BreedListScreen(
+                allTemperaments = allTemperaments,
+                onDismiss = { handleSheetClose(false) },
+                onDirtyChange = { isSheetDirty.value = it },
+                onSave = { name, origin, desc, temps, min, max, imageUrl ->
```
Maybe send the Breed class already. This is not type safe and not very scalable, imagine that now we add 5 new params to the breed, this would get big and hard to maintain

AlexCarvalhoSword commented on features/breed_list/src/main/java/com/example/catbreeds/breed_list/NewBreedSheet.kt:
```diff
@@ -0,0 +1,682 @@
+    val context = LocalContext.current
+
+    // Field Value States
+    val name = remember { mutableStateOf("") }
```
we could have a StateHolder class with all of these values. There you will also have all the validations. This way you centralize all the logic in a single class and make the validations testable. 

this would be something like: 

```
FormState() {
    private val _origin = mutableStateOf("")
    val origin: State<String> = _origin //    (or use the breed class already )
    // (....)
   
    fun validate() // (...)
    
    } 
   
```   

 Then you can create the method with a remember 
 
` fun rememberFormState() = remember { FormState() }` 


This is a similar approach to what, for example, `LazyColumn` has with its `LazyListState`

goncalo-guilherme-sword commented on features/breed_list/src/main/java/com/example/catbreeds/breed_list/NewBreedSheet.kt:
```diff
@@ -0,0 +1,682 @@
+    allTemperaments: List<String>,
+    onDismiss: () -> Unit,
+    onDirtyChange: (Boolean) -> Unit,
+    onSave: (String, String, String, List<String>, String, String, String) -> Unit
```
As we spoke, and since this is the callback called with the new breed informations, you can replace this with the Breed data class.

AlexCarvalhoSword commented on features/breed_list/src/main/java/com/example/catbreeds/breed_list/NewBreedSheet.kt:
```diff
@@ -0,0 +1,682 @@
+@Composable
+private fun LifeExpectancyField(
+    minLife: String,
+    onMinLifeChange: (String) -> Unit,
```
just an heads up: on our project we usually keep lambdas as last params

gfrantzsword committed:
refactor: simplify `addNewBreed` to accept `Breed` object

- Added new `createBreed()` method on `NewBreedSheet.kt` to improve readability

gfrantzsword committed:
refactor: Removed redundant `refreshBreeds()` on BreedListViewModel's `init`

- Removed obsolete unit test

gfrantzsword committed:
refactor: Made all lambdas on `NewBreedSheet.kt` the last parameter(s)

gfrantzsword committed:
refactor: Created `NewBreedFormState` stateHolder class to centralize Field State logic

gfrantzsword committed:
refactor: Created `NewBreedViewModel` class to separate from original `BreedListViewModel`

gfrantzsword committed:
refactor: Pass `onSuccess` as a callback to the viewmodel

gfrantzsword committed:
test: Created `NewBreedViewModelTest.kt`

gfrantzsword committed:
refactor: Update readme

gfrantzsword committed:
Revert "refactor: Update readme"

This reverts commit 38be704a2910c103301f499768903d322d31b8af.

gfrantzsword committed:
fix: Search field not on focus when opened

gfrantzsword committed:
test: Add NewBreedFormStateTest to test form fields and validations

gfrantzsword committed:
refactor: Create strings.xml and update all instances

gfrantzsword committed:
test: Update form state test to use `strings.xml` constants

gfrantzsword committed:
feature: Add 'delete' feature to custom breeds + small changes

- text fields loose focus when a temperament chip is selected
- `Add new breed` submit button now transitions into a smaller (non-extended) version when the keyboard is active
- Moved more instances of strings into ´´strings.xml

AlexCarvalhoSword commented on features/breed_list/src/main/java/com/example/catbreeds/breed_list/NewBreedViewModel.kt:
```diff
@@ -0,0 +1,71 @@
+
+    private val _errorMessage = mutableStateOf<String?>(null)
+
+    val newIdPrefix = R.string.new_breed_id_prefix.toString()
```
wouldn't it be better to have a new property on the Breed itself that would tell us if the breed is a custom one ?

AlexCarvalhoSword commented on features/breed_list/src/main/java/com/example/catbreeds/breed_list/NewBreedViewModel.kt:
```diff
@@ -0,0 +1,71 @@
+    @param:ApplicationContext private val context: Context
+) : ViewModel() {
+
+    private val _errorMessage = mutableStateOf<String?>(null)
```
not used I think

gfrantzsword committed:
fix: Fix form state reset on submission + remove 'temperaments' requirements

gfrantzsword committed:
refactor: Add `isCustom` property to new breeds

goncalo-guilherme-sword commented on data/src/main/java/com/example/catbreeds/data/repository/BreedRepositoryImpl.kt:
```diff
@@ -92,4 +93,28 @@ class BreedRepositoryImpl(
     }
+
+    override suspend fun addBreed(breed: Breed) {
+        val entity = BreedEntity(
```
Instead of constructing this object here, you could have created a mapper for this. It would make the code cleaner, nicer, and adds the possibility of reusing in other places. 
Something like this:

`fun Breed.toBreedEntity() = BreedEntity(...)`

And then call this mapper here.

goncalo-guilherme-sword commented on features/breed_list/src/main/java/com/example/catbreeds/breed_list/BreedListScreen.kt:
```diff
@@ -77,12 +152,73 @@ fun BreedListScreen(
+                allTemperaments = allTemperaments,
+                onDismiss = { handleSheetClose(false) },
+                onDirtyChange = { isSheetDirty.value = it },
+                onSave = { breed ->
```
Since you are only using the NewBreedViewModel here, I believe you could move this ViewModel to inside the BottomSheet that you use for the breeds `NewBreedSheetContent` and call the save method from the ViewModel directly there.
